zenith_utils/src/nonblock.rs

```diff
@@ -0,0 +1,19 @@
+use nix::fcntl::{fcntl, OFlag, F_GETFL, F_SETFL};
+use std::os::unix::io::RawFd;
+
+/// Put a file descriptor into non-blocking mode
+pub fn set_nonblock(fd: RawFd) -> Result<(), std::io::Error> {
+    let bits = fcntl(fd, F_GETFL)?;
+
+    // Safety: If F_GETFL returns some unknown bits, they should be valid
+    // for passing back to F_SETFL, too. If we left them out, the F_SETFL
+    // would effectively clear them, which is not what we want.
+    let mut flags = unsafe {
+        OFlag::from_bits_unchecked(bits)
+    };
```

Comment:
Is there an advantage to this over from_bits/from_bits_truncate?

Reply:
`from_bits` returns None if there are any unknown bits. We'd have to turn that into a panic or other error.

`from_bits_truncate` would silently zero out any unknown bits. That's bad, because then when we call F_SETFL to set the O_NONBLOCK flag, we will clear all those unknown bits as a side effect. It's hard to tell if clearing the unknown bits would be OK or not without knowing what they are.

So from_bits_unchecked() seems like the safest option. It will leave any unknown bits untouched, which is what we want. We just want to set O_NONBLOCK, we don't care what existing flags were set.

This is all pretty hypothetical, introducing new file status flags is rare, and even more rare that they would be set by default on the process pipes. Nevertheless, `from_bits_unchecked` seems like the best option to me, although you could also make an argument for `from_bits` and panicing; then it would fail loudly if some new flags are being introduced and we'd be forced to make a decision on them. But that's a bit wishy washy too, we would only hit that panic if a new flag is introduced that 'nix' doesn't know about. We still wouldn't notice if a new flag is introduced in the kernel, and in 'nix'. So that seems a bit pointless, too.

Reply:
Ahh ok this seems fine, the unsafe made me uneasy but I don't think it's a real concern.